File: cpp/countingSort/countingSort.cpp
#include <iostream>
#include <climits>

int* countingSort(int* array, int len) {
	int maxElement = INT_MIN;
	for (int i = 0; i < len; ++i) {
		if (array[i] > maxElement) {
			maxElement = array[i];
		}
	}

	int* countArray = new int[maxElement+1]{};
	
	for (int i = 0; i < len; ++i) {
		countArray[array[i]]++;
	}

	for (int i = 1; i < maxElement+1; ++i) {
		countArray[i] += countArray[i-1];
	}

	int* resultArray = new int[len]{};
	for (int i = len-1; i >= 0; --i) {
		resultArray[countArray[array[i]] -1] = array[i];
		countArray[array[i]]--;
	}
	
	delete[] countArray;

	return resultArray;
}

void printArray(int* array, int len) {
	for (int i = 0; i < len; ++i) {
		std::cout << array[i] << " ";
	}
	std::cout << "\n";
}

int main() {
	int* array = new int[]{12, 11, 13, 5, 6};
	int len = 5;

	int* result = countingSort(array, len);
	printArray(result, len);

	delete[] array;
	delete[] result;
	return 0;
}

File: python/countingSort/countingSort.py

def counting_sort(l):
    maxElement = float('-inf')
    for e in l:
        if e > maxElement: 
            maxElement = e

    countingList = [0 for i in range(maxElement+1)]
    for e in l:
        countingList[e] += 1

    for i in range(1, len(countingList)):
        countingList[i] += countingList[i-1]

    result = [0 for i in range(len(l))]
    for i in range(len(l)-1, -1, -1):
        result[countingList[l[i]]-1] = l[i]
    return result


l = [12, 11, 13, 5, 6]
result = counting_sort(l)
print(result)
